File: frontend/src/runtime/streamEvents.ts
import type { PermissionRequest, RuntimeEvent, RuntimeStreamEvent } from "../api/schemas.ts";
import type { ActivityItem } from "./reducer.ts";

export function runtimeStreamEvent(event: RuntimeEvent): RuntimeStreamEvent | null {
  const payload = recordValue(event.data.stream_event);
  const kind = stringValue(payload.kind);
  if (!kind) return null;
  return payload as RuntimeStreamEvent;
}

export function activityFromStreamEvent(event: RuntimeEvent, streamEvent: RuntimeStreamEvent): ActivityItem | null {
  if (streamEvent.kind === "assistant_delta" || streamEvent.kind === "assistant_final") {
    return null;
  }
  if (streamEvent.kind === "progress") {
    return {
      id: event.id,
      kind: "runtime",
      label: streamEvent.message,
      summary: "",
      status: "info",
      timestamp: event.timestamp,
      eventType: "stream.progress",
      category: "runtime",
      data: {
        stream_event_kind: streamEvent.kind,
        message: streamEvent.message,
        stage: streamEvent.stage ?? null,
        message_id: streamEvent.message_id ?? null,
      },
    };
  }
  if (streamEvent.kind === "tool_lifecycle") {
    const status = toolLifecycleStatus(streamEvent.phase);
    const details = recordValue(streamEvent.details);
    return {
      id: `tool:${streamEvent.tool_call_id}:${streamEvent.phase}`,
      kind: "tool",
      label: streamEvent.title || streamEvent.tool_name,
      summary: streamEvent.result_summary || streamEvent.args_summary || streamEvent.error?.message || "",
      status,
      timestamp: event.timestamp,
      eventType: "stream.tool_lifecycle",
      category: "tool",
      data: {
        ...details,
        stream_event_kind: streamEvent.kind,
        phase: streamEvent.phase,
        title: streamEvent.title ?? null,
        tool_call_id: streamEvent.tool_call_id,
        tool_name: streamEvent.tool_name,
        args_summary: streamEvent.args_summary ?? null,
        result_summary: streamEvent.result_summary ?? null,
        command: streamEvent.command ?? details.command,
        path: streamEvent.path ?? details.path,
        exit_code: streamEvent.exit_code ?? details.exit_code,
        duration_ms: streamEvent.duration_ms ?? details.duration_ms,
        error: streamEvent.error ?? null,
      },
    };
  }
  if (streamEvent.kind === "permission_state") {
    return {
      id: `permission:${streamEvent.tool_call_id}:${streamEvent.status}`,
      kind: "permission",
      label: permissionLabel(streamEvent),
      summary: streamEvent.reason || streamEvent.args_summary || "",
      status: permissionStatus(streamEvent.status),
      timestamp: event.timestamp,
      eventType: "stream.permission_state",
      category: "permission",
      data: {
        stream_event_kind: streamEvent.kind,
        status: streamEvent.status,
        tool_call_id: streamEvent.tool_call_id,
        tool_name: streamEvent.tool_name,
        action: streamEvent.action ?? null,
        risk: streamEvent.risk ?? null,
        args_summary: streamEvent.args_summary ?? null,
        reason: streamEvent.reason ?? null,
        args: streamEvent.args ?? {},
      },
    };
  }
  if (streamEvent.kind === "error") {
    return {
      id: event.id,
      kind: "error",
      label: streamEvent.error_type || "Runtime error",
      summary: streamEvent.message,
      status: "error",
      timestamp: event.timestamp,
      eventType: "stream.error",
      category: "error",
      data: { stream_event_kind: streamEvent.kind, ...streamEvent },
    };
  }
  if (streamEvent.kind === "artifact") {
    return {
      id: event.id,
      kind: "event",
      label: streamEvent.title || streamEvent.artifact_kind,
      summary: streamEvent.uri || "",
      status: "success",
      timestamp: event.timestamp,
      eventType: "stream.artifact",
      category: "artifact",
      data: { stream_event_kind: streamEvent.kind, ...streamEvent },
    };
  }
  return null;
}

export function permissionRequestFromStreamEvent(streamEvent: RuntimeStreamEvent): PermissionRequest | null {
  if (streamEvent.kind !== "permission_state" || streamEvent.status !== "required") {
    return null;
  }
  return {
    tool_call_id: streamEvent.tool_call_id,
    tool_name: streamEvent.tool_name,
    action: streamEvent.action ?? null,
    risk: streamEvent.risk ?? null,
    args_summary: streamEvent.args_summary ?? null,
    reason: streamEvent.reason ?? null,
    args: streamEvent.args ?? {},
  };
}

function toolLifecycleStatus(phase: Extract<RuntimeStreamEvent, { kind: "tool_lifecycle" }>["phase"]): ActivityItem["status"] {
  if (phase === "started" || phase === "scheduled" || phase === "permission_required") return "running";
  if (phase === "completed") return "success";
  if (phase === "failed") return "error";
  if (phase === "blocked") return "blocked";
  return "info";
}

function permissionStatus(status: Extract<RuntimeStreamEvent, { kind: "permission_state" }>["status"]): ActivityItem["status"] {
  if (status === "required") return "pending";
  if (status === "approved") return "success";
  if (status === "rejected" || status === "blocked") return "blocked";
  return "info";
}

function permissionLabel(streamEvent: Extract<RuntimeStreamEvent, { kind: "permission_state" }>): string {
  if (streamEvent.status === "required") return "Permission required";
  if (streamEvent.status === "approved") return "Permission approved";
  if (streamEvent.status === "rejected") return "Permission rejected";
  return "Permission blocked";
}

function recordValue(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : {};
}

function stringValue(value: unknown): string {
  return typeof value === "string" && value ? value : "";
}

File: src/langgraph_agent_blueprint/models/stream_events.py
"""Typed public stream-event payloads carried by RuntimeEvent.data['stream_event']."""

from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import Field, TypeAdapter

from .base import FrozenRuntimeModel


StreamEventKind = Literal[
    "assistant_delta",
    "assistant_final",
    "progress",
    "tool_lifecycle",
    "permission_state",
    "subagent",
    "error",
    "artifact",
]
ToolLifecyclePhase = Literal["scheduled", "started", "completed", "failed", "blocked", "permission_required"]
PermissionStateStatus = Literal["required", "approved", "rejected", "blocked"]
SubagentPhase = Literal["started", "event", "finished", "error", "cancelled", "timeout"]


class AssistantDeltaStreamEvent(FrozenRuntimeModel):
    """One public assistant text delta for a visible assistant message."""

    kind: Literal["assistant_delta"] = "assistant_delta"
    message_id: str
    delta: str


class AssistantFinalStreamEvent(FrozenRuntimeModel):
    """Final visible assistant answer for a turn."""

    kind: Literal["assistant_final"] = "assistant_final"
    message_id: str
    content: str


class ProgressStreamEvent(FrozenRuntimeModel):
    """Concise public progress narration, not hidden reasoning."""

    kind: Literal["progress"] = "progress"
    message: str
    stage: str | None = None
    message_id: str | None = None


class StreamError(FrozenRuntimeModel):
    """Normalized public error summary."""

    type: str
    message: str


class ToolLifecycleStreamEvent(FrozenRuntimeModel):
    """One lifecycle transition for a model-requested tool call."""

    kind: Literal["tool_lifecycle"] = "tool_lifecycle"
    phase: ToolLifecyclePhase
    tool_call_id: str
    tool_name: str
    title: str | None = None
    args_summary: str | None = None
    result_summary: str | None = None
    command: str | None = None
    path: str | None = None
    exit_code: int | None = None
    duration_ms: float | None = None
    error: StreamError | None = None
    details: dict[str, Any] = Field(default_factory=dict)


class PermissionStateStreamEvent(FrozenRuntimeModel):
    """Human-in-the-loop or policy state for one tool call."""

    kind: Literal["permission_state"] = "permission_state"
    status: PermissionStateStatus
    tool_call_id: str
    tool_name: str
    action: str | None = None
    risk: str | None = None
    args_summary: str | None = None
    reason: str | None = None
    args: dict[str, Any] = Field(default_factory=dict)
    scope: Literal["tool", "subagent"] = "tool"
    parent_session_id: str | None = None
    parent_thread_id: str | None = None
    child_session_id: str | None = None
    child_thread_id: str | None = None
    child_run_id: str | None = None
    subagent_name: str | None = None


class SubagentStreamEvent(FrozenRuntimeModel):
    """Typed parent-stream envelope for one subagent lifecycle or child event."""

    kind: Literal["subagent"] = "subagent"
    phase: SubagentPhase
    subagent_id: str
    run_id: str
    sequence: int
    parent_session_id: str | None = None
    parent_thread_id: str | None = None
    child_session_id: str | None = None
    child_thread_id: str | None = None
    agent_call_id: str | None = None
    name: str | None = None
    purpose: str | None = None
    status: str | None = None
    summary: str | None = None
    child_event_id: str | None = None
    child_event_type: str | None = None
    child_event: dict[str, Any] = Field(default_factory=dict)
    child_stream_event: dict[str, Any] = Field(default_factory=dict)
    error: StreamError | None = None


class ErrorStreamEvent(FrozenRuntimeModel):
    """Runtime error event rendered from typed data rather than event-name heuristics."""

    kind: Literal["error"] = "error"
    message: str
    error_type: str | None = None
    recoverable: bool | None = None


class ArtifactStreamEvent(FrozenRuntimeModel):
    """Artifact reference created during a runtime turn."""

    kind: Literal["artifact"] = "artifact"
    artifact_id: str
    artifact_kind: str
    uri: str | None = None
    title: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


RuntimeStreamEvent = Annotated[
    AssistantDeltaStreamEvent
    | AssistantFinalStreamEvent
    | ProgressStreamEvent
    | ToolLifecycleStreamEvent
    | PermissionStateStreamEvent
    | SubagentStreamEvent
    | ErrorStreamEvent
    | ArtifactStreamEvent,
    Field(discriminator="kind"),
]

_RUNTIME_STREAM_EVENT_ADAPTER = TypeAdapter(RuntimeStreamEvent)


def stream_event_payload(payload: RuntimeStreamEvent | dict[str, Any]) -> dict[str, Any]:
    """Validate and dump a public stream-event payload."""

    return _RUNTIME_STREAM_EVENT_ADAPTER.validate_python(payload).model_dump(mode="json", exclude_none=True)
